lib/helpers.py

```diff
@@ -10,437 +10,598 @@
+            # s3 cache (useful if we don't have enough space on the Colab instance)
+            s3_cache_file_name = '{0}/{1}/{2}/{3}-{4}.parquet' \
+                .format(self._audience_data_path(audience), source, cache_folder, date_str, self.revenue_event)
 
-    print(datetime.now(), 'finished loading CSV for', date.strftime('%d.%m.%Y'),
-          'took', datetime.now() - now)
+        fs = s3fs.S3FileSystem(anon=False)
+        fs.connect_timeout = 10  # defaults to 5
+        fs.read_timeout = 30  # defaults to 15
 
-    if not os.path.exists(cache_dir):
-        os.makedirs(cache_dir)
+        columns = self.columns.get(source)
 
-    print(datetime.now(), 'caching local as parquet', cache_filename)
-    to_parquet(df, cache_filename)
+        if os.path.exists(cache_file_name):
+            print(now, 'loading from', cache_file_name)
+            return self._from_parquet_corrected(
+                file_name=cache_file_name,
+                s3_file_name=s3_cache_file_name,
+                fs=fs,
+                columns=columns,
+            )
 
-    # write it to the S3 cache folder as well
-    print(datetime.now(), 'caching on S3 as parquet', s3_cache_filename)
-    to_parquet(df, s3_cache_filename)
+        if fs.exists(path=s3_cache_file_name):
+            print(now, 'loading from S3 cache', s3_cache_file_name)
 
-    return df
+            # Download the file to local cache first to avoid timeouts during the load.
+            # This way, if they happen, restart will be using local copies first.
+            fs.get(s3_cache_file_name, cache_file_name)
 
+            print(now, 'stored S3 cache file to local drive, loading', cache_file_name)
 
-def calculate_ad_spend(df):
-    ad_spend_micros = df[(df.event_type == 'buying_conversion') & (df.ab_test_group == TEST)]['cost_eur'].sum()
-    return ad_spend_micros / 10 ** 6
+            return self._from_parquet_corrected(
+                file_name=cache_file_name,
+                s3_file_name=s3_cache_file_name,
+                fs=fs,
+                columns=columns,
+            )
 
+        print(now, 'start loading CSV for', audience, source, date)
 
-def marked(df):
-    """
-    The dataframe created by `marked` will contain all mark events. Remerge marks users per campaign. If a user was
-    marked once for an audience he will have the same group allocation for consecutive marks (different campaigns)
-    unless manually reset on audience level.
-    """
-    mark_df = df[df.event_type == 'mark']
+        read_csv_kwargs = {'chunksize': self.chunk_size}
+        if columns:
+            read_csv_kwargs['usecols'] = columns
 
-    # we dont need the event_type anymore (to save memory)
-    mark_df = mark_df.drop(columns=['event_type'])
+        df = pd.DataFrame()
 
-    sorted_mark_df = mark_df.sort_values('ts')
+        if not fs.exists(path=filename):
+            print(now, 'WARNING: no CSV file at for: ', audience, source, date, ', skipping the file: ', filename)
```

Comment:
It could be too much. Wouldn't it better to use the logging module to differentiate warnings from info messages?

Reply:
No, logging module produces pretty weird looking outputs in notebooks - say, in default jupyter it's always scary red.

This is why I defined a `log` function in the class... and didn't use it, will fix that.

Reply:
Additionally, we don't have different logging levels here now at all, when we would need to differentiate, then we can switch over to using the `logging` module, of course.

Reply:
I thought we have since there's explicit `WARNING:` in the message, but of course, it doesn't worth to use `logging` if it's producing any weird stuff into notebooks.